Errors occur when reading a value from the store implementation for bbolt / Bolt DB

The recent Travis CI builds have some errors:
Example 1 from https://travis-ci.org/philippgille/gokv/builds/445281693:
```
=== RUN   TestStoreConcurrent
--- FAIL: TestStoreConcurrent (3.38s)
	test.go:74: invalid character '\x00' looking for beginning of value
	test.go:74: invalid character '\x00' looking for beginning of value
	test.go:74: invalid character '\x00' looking for beginning of value
FAIL
coverage: 81.1% of statements
FAIL	github.com/philippgille/gokv/bolt	3.420s
```
Example 2 from https://travis-ci.org/philippgille/gokv/builds/445303063:
```
=== RUN   TestStoreConcurrent
unexpected fault address 0x7f361ca600ef
fatal error: fault
[signal SIGSEGV: segmentation violation code=0x1 addr=0x7f361ca600ef pc=0x5c2444]
goroutine 25 [running]:
runtime.throw(0x627b33, 0x5)
	/home/travis/.gimme/versions/go1.10.linux.amd64/src/runtime/panic.go:619 +0x81 fp=0xc420053c48 sp=0xc420053c28 pc=0x453861
runtime.sigpanic()
	/home/travis/.gimme/versions/go1.10.linux.amd64/src/runtime/signal_unix.go:395 +0x211 fp=0xc420053c98 sp=0xc420053c48 pc=0x468cf1
encoding/json.checkValid(0x7f361ca600ef, 0xa, 0xa, 0xc420192260, 0x631b60, 0xc420053d70)
	/home/travis/.gimme/versions/go1.10.linux.amd64/src/encoding/json/scanner.go:27 +0x144 fp=0xc420053d00 sp=0xc420053c98 pc=0x5c2444
encoding/json.Unmarshal(0x7f361ca600ef, 0xa, 0xa, 0x5e34e0, 0xc42002c860, 0xc4200c5500, 0x0)
	/home/travis/.gimme/versions/go1.10.linux.amd64/src/encoding/json/decode.go:102 +0xbd fp=0xc420053d70 sp=0xc420053d00 pc=0x5addfd
github.com/philippgille/gokv/util.FromJSON(0x7f361ca600ef, 0xa, 0xa, 0x5e34e0, 0xc42002c860, 0x0, 0xc42002c860)
	/home/travis/gopath/src/github.com/philippgille/gokv/util/util.go:12 +0x65 fp=0xc420053dc8 sp=0xc420053d70 pc=0x5c8e25
github.com/philippgille/gokv/bolt.Store.Get(0xc4200e0000, 0x627f2d, 0x7, 0x631832, 0x1, 0x5e34e0, 0xc42002c860, 0xc420053ee8, 0x437b08, 0x10)
	/home/travis/gopath/src/github.com/philippgille/gokv/bolt/bolt.go:55 +0x203 fp=0xc420053e80 sp=0xc420053dc8 pc=0x5c9423
github.com/philippgille/gokv/bolt.(*Store).Get(0xc4200f21e0, 0x631832, 0x1, 0x5e34e0, 0xc42002c860, 0x0, 0x0, 0x0)
	<autogenerated>:1 +0xb3 fp=0xc420053ef8 sp=0xc420053e80 pc=0x5ca403
github.com/philippgille/gokv/test.InteractWithStore(0x646980, 0xc4200f21e0, 0x631832, 0x1, 0xc4200d6000, 0xc4200da130)
	/home/travis/gopath/src/github.com/philippgille/gokv/test/test.go:72 +0x297 fp=0xc420053fb0 sp=0xc420053ef8 pc=0x5cc9a7
runtime.goexit()
	/home/travis/.gimme/versions/go1.10.linux.amd64/src/runtime/asm_amd64.s:2361 +0x1 fp=0xc420053fb8 sp=0xc420053fb0 pc=0x4837f1
created by github.com/philippgille/gokv/bolt_test.TestStoreConcurrent
	/home/travis/gopath/src/github.com/philippgille/gokv/bolt/bolt_test.go:44 +0x241

[...]

goroutine 244 [runnable]:
github.com/philippgille/gokv/test.InteractWithStore(0x646980, 0xc42000de20, 0xc420018bc9, 0x3, 0xc4200d6000, 0xc4200da130)
	/home/travis/gopath/src/github.com/philippgille/gokv/test/test.go:58
created by github.com/philippgille/gokv/bolt_test.TestStoreConcurrent
	/home/travis/gopath/src/github.com/philippgille/gokv/bolt/bolt_test.go:44 +0x241
FAIL	github.com/philippgille/gokv/bolt	0.089s
```
Example 3 from https://travis-ci.org/philippgille/gokv/builds/445310684:
```
=== RUN   TestStoreConcurrent
--- FAIL: TestStoreConcurrent (2.76s)
	test.go:74: invalid character '\x01' looking for beginning of value
FAIL
coverage: 81.1% of statements
FAIL	github.com/philippgille/gokv/bolt	2.791s
```

While reading about [BadgerDB](https://github.com/dgraph-io/badger) as one of the next `gokv.Store` implementations I read their warning about the validity of data from within transactions, which is:

> Please note that values returned from `Get()` are only valid while the transaction is open. If you need to use a value outside of the transaction then you must use `copy()` to copy it to another byte slice.

And I remembered that I read something similar regarding bbolt, which is written in their GoDocs: https://godoc.org/go.etcd.io/bbolt#hdr-Caveats (or to have a future working link, [here](https://github.com/etcd-io/bbolt/blob/76a4670663d125b6b89d47ea3cc659a282d87c28/doc.go#L38).

To quote:

> Keys and values retrieved from the database are only valid for the life of the transaction. When used outside the transaction, these byte slices can point to different data or can point to invalid memory which will cause a panic.

Yet, the current implementation in `gokv` doesn't take this into account:
```
var data []byte
c.db.View(func(tx *bolt.Tx) error {
	b := tx.Bucket([]byte(c.bucketName))
	data = b.Get([]byte(k))
	// [...]
	return nil
})
// ... continue to work with data
```

So this could very well be the reason for the errors we see in the Travis CI log.

This needs to be fixed as soon as possible!

--- a/RELEASES.md
+++ b/RELEASES.md
@@ -11,6 +11,7 @@ vNext
 - Added: Package `gomap` - A `gokv.Store` implementation for a plain Go map with a `sync.RWMutex` for concurrent access (issue [#11](https://github.com/philippgille/gokv/issues/11))
 - Improved: Every `gokv.Store` implementation resides in its own package now, so when downloading the package of an implementation, for example with `go get github.com/philippgille/gokv/redis`, only the actually required dependencies are downloaded and compiled, making the process much faster. This is especially useful for example when creating Docker images, where in many cases (depending on the `Dockerfile`) the download and compilation are repeated for *each build*. (Issue [#2](https://github.com/philippgille/gokv/issues/2))
 - Improved: The performance of BoltClient should be higher, because unnecessary manual locking was removed. (Issue [#1](https://github.com/philippgille/gokv/issues/1))
+- Fixed: The `gokv.Store` implementation for bbolt / Bolt DB used data from within a Bolt transaction outside of it, without copying the value, which can lead to errors (see [here](https://github.com/etcd-io/bbolt/blob/76a4670663d125b6b89d47ea3cc659a282d87c28/doc.go#L38)) (issue [#13](https://github.com/philippgille/gokv/issues/13))
 
 ### Breaking changes
 

--- a/bolt/bolt.go
+++ b/bolt/bolt.go
@@ -23,8 +23,7 @@ func (c Store) Set(k string, v interface{}) error {
 
 	err = c.db.Update(func(tx *bolt.Tx) error {
 		b := tx.Bucket([]byte(c.bucketName))
-		err := b.Put([]byte(k), data)
-		return err
+		return b.Put([]byte(k), data)
 	})
 	if err != nil {
 		return err
@@ -38,14 +37,20 @@ func (c Store) Set(k string, v interface{}) error {
 // that v points to with the values of the retrieved object's values.
 func (c Store) Get(k string, v interface{}) (bool, error) {
 	var data []byte
-	err := c.db.View(func(tx *bolt.Tx) error {
+	c.db.View(func(tx *bolt.Tx) error {
 		b := tx.Bucket([]byte(c.bucketName))
-		data = b.Get([]byte(k))
+		txData := b.Get([]byte(k))
+		// txData is only valid during the transaction.
+		// Its value must be copied to make it valid outside of the tx.
+		// TODO: Benchmark if it's faster to copy + close tx,
+		// or to keep the tx open until unmarshalling is done.
+		if txData != nil {
+			// `data = append([]byte{}, txData...)` would also work, but the following is more explicit
+			data = make([]byte, len(txData))
+			copy(data, txData)
+		}
 		return nil
 	})
-	if err != nil {
-		return false, err
-	}
 
 	// If no value was found assign nil to the pointer
 	if data == nil {

